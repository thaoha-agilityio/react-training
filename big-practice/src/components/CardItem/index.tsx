import { memo } from 'react';
import {
  Card,
  CardBody,
  CardFooter,
  Flex,
  Image,
  Heading,
  Text,
  Stack,
  IconButton,
  useDisclosure,
  Button,
  Box,
} from '@chakra-ui/react';
import { DeleteIcon, EditIcon, InfoIcon } from '@chakra-ui/icons';
import ConfirmModal from '@components/ConfirmModal';

// Types
import { IProduct } from '@types';

type CardItemProps = {
  item: IProduct;
  isLoading?: boolean;
  onDeleteItem: () => void;
  onEditItem: () => void;
  onShowDetailItem: () => void;
  onAddToCart: () => void;
};

export const CardItem = memo(
  ({ item, onAddToCart, onDeleteItem, onEditItem, onShowDetailItem, isLoading }: CardItemProps) => {
    const { name, description, price, image } = item;
    const { isOpen, onOpen, onClose } = useDisclosure();

    return (
      <Flex pos='relative'>
        <Card>
          <CardBody role='group'>
            <Image src={image} alt='card-item' />
            <Box
              zIndex={1}
              pos='absolute'
              bgColor='gray.250'
              opacity={0.72}
              top={0}
              right={0}
              left={0}
              bottom={0}
              _groupHover={{ display: 'block' }}
              display='none'
            />
            <Flex
              zIndex={9}
              pos='absolute'
              top='160px'
              right='20px'
              _groupHover={{ display: 'flex' }}
              display='none'
              flexDir='column'
              alignItems='center'
              gap='10px'
            >
              <Button w='202px' h='48px' onClick={onAddToCart}>
                Add to cart
              </Button>

              <Flex gap='35px' zIndex={50} justifyContent='center'>
                <Flex alignItems='baseline'>
                  <IconButton
                    variant='unstyled'
                    aria-label='delete-card'
                    data-testid='delete-btn'
                    icon={<DeleteIcon />}
                    rounded='xs'
                    onClick={onOpen}
                    color='white'
                  />
                  <Text color='white'>Delete</Text>

                  <IconButton
                    aria-label='edit-card'
                    data-testid='edit-btn'
                    icon={<EditIcon />}
                    rounded='xs'
                    variant='unstyled'
                    onClick={onEditItem}
                    color='white'
                  />
                  <Text color='white'>Edit</Text>

                  <IconButton
                    aria-label='info-card'
                    data-testid='detail-btn'
                    icon={<InfoIcon />}
                    rounded='xs'
                    variant='unstyled'
                    onClick={onShowDetailItem}
                    color='white'
                  />
                  <Text color='white'>Detail</Text>
                </Flex>
              </Flex>
            </Flex>
          </CardBody>
          <CardFooter>
            <Stack spacing='8px' maxW={{ base: '160px', md: '260px' }}>
              <Heading fontSize={{ base: 'base', md: 'lg' }}>{name}</Heading>
              <Text
                color='gray.150'
                fontWeight='medium'
                overflow='hidden'
                whiteSpace='nowrap'
                textOverflow='ellipsis'
              >
                {description}
              </Text>
              <Text fontSize={{ base: 'tiny', md: 'md' }}>Rp {price}</Text>
            </Stack>
          </CardFooter>
        </Card>
        <ConfirmModal
          isOpen={isOpen}
          title='Delete Confirmation'
          textCancel='Cancel'
          textSubmit='Yes, Delete'
          text='Are you sure you want to delete this item?'
          onClose={onClose}
          onSubmit={onDeleteItem}
          isLoading={isLoading}
        />
      </Flex>
    );
  },
);
