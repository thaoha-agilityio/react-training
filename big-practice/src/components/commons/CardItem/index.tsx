import { Card, CardBody, CardFooter, Flex, Image, Heading, Text, Stack } from '@chakra-ui/react';
import { IProduct } from '@types';

type CardItemProps = {
  item: IProduct;
};

export const CardItem = ({ item: { name, description, price, image } }: CardItemProps) => {
  return (
    <Flex>
      <Card>
        <CardBody>
          <Image src={image} alt='card-item' />
        </CardBody>
        <CardFooter>
          <Stack spacing='8px' maxW={{ base: '160px', md: '260px' }}>
            <Heading fontSize={{ base: 'base', md: 'lg' }}>{name}</Heading>
            <Text
              color='gray.150'
              fontWeight='medium'
              overflow='hidden'
              whiteSpace='nowrap'
              textOverflow='ellipsis'
            >
              {description}
            </Text>
            <Text fontSize={{ base: 'tiny', md: 'md' }}>Rp {price}</Text>
          </Stack>
        </CardFooter>
      </Card>
    </Flex>
  );
};
