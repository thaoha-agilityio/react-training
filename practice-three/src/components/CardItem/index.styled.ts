import styled from "styled-components";

import { ColumnBaselineStyle } from "../../styled-common/layout";
import { colors, fonts, metrics } from "../../themes";
import { P } from "../../styled-common/paragraph";
import { fontSizes, lineHeights } from "../../themes/fonts";

type CardItemProps = {
  isGridView: boolean;
};

const CardItemStyled = styled.div<CardItemProps>`
  width: ${(props) =>
    props.isGridView ? `${metrics.widths.lg}px` : `${metrics.widths["3xl"]}px`};
  height: ${(props) =>
    props.isGridView
      ? `${metrics.heights["4xl"]}px`
      : `${metrics.heights.xl}px`};
  box-shadow: ${metrics.shadows.sm};
  border-radius: ${metrics.borderRadius.tiny}px;
  display:flex;
  align-items:${(props) => props.isGridView ? 'baseline' : 'center'};
  background-color: ${({ theme }) => theme.mainHeading};
  flex-direction: ${(props) => (props.isGridView ? " column" : "row")};
  cursor: pointer;
`;

const CardImgWrapperStyled = styled.div<CardItemProps>`
  width: ${(props) => (props.isGridView ? `249px` : "396px")};
  height: ${(props) => (props.isGridView ? `238px` : "178px")};
  background-color: ${colors.linkWater};
  border-radius: ${metrics.borderRadius.tiny}px;
  margin: 13px 18px;
  padding: 18px 20px;
`;

const CardContentStyled = styled.div`
  margin: 21px 19px;
  font-weight: ${fonts.fontWeights.semiBold};
`;

const CardTitleStyled = styled(P.Medium) <CardItemProps>`
  color: ${({ theme }) => theme.textColor};
  ${(props) => !props.isGridView && `font-size: ${fontSizes.xxl}px;`}
  ${(props) => !props.isGridView && `line-height: ${lineHeights.xl};`}
`;

const CardInfoStyled = styled(P.Small) <CardItemProps>`
  color: ${({ theme }) => theme.infoBook};
  ${(props) => !props.isGridView && `font-size: ${fontSizes.lg}px;`}
  ${(props) => !props.isGridView && `line-height: ${lineHeights.sm};`}
`;

export {
  CardItemStyled,
  CardImgWrapperStyled,
  CardInfoStyled,
  CardContentStyled,
  CardTitleStyled,
};
