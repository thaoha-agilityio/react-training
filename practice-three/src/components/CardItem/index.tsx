import { memo, useCallback } from "react";

import Avatar from "../Avatar";

import { IBook } from "../../types/book";

import {
  CardItemStyled,
  CardImgWrapperStyled,
  CardContentStyled,
  CardInfoStyled,
  CardTitleStyled,
} from "./index.styled";

type CardItemProps = {
  isGridView: boolean;
  item: IBook;
  onSetSelectedBookId: (id: string) => void;
}

const CardItem = ({
  item: { id, avatar, name, author, published },
  onSetSelectedBookId,
  isGridView,
}: CardItemProps): React.ReactElement => {
  const handleClick = useCallback((): void => {
    onSetSelectedBookId(id);
  }, []);

  return (
    <CardItemStyled
      data-testid="card-item"
      onClick={handleClick}
      isGridView={isGridView}
    >
      <CardImgWrapperStyled isGridView={isGridView}>
        <Avatar
          url={avatar}
          alt="avatar"
          borderRadius={5}
          width={isGridView ? 201 : 353}
          height={isGridView ? 202 : 144}
        />
      </CardImgWrapperStyled>
      <CardContentStyled>
        <CardTitleStyled isGridView={isGridView}>{name}</CardTitleStyled>
        <CardInfoStyled isGridView={isGridView}>{author}</CardInfoStyled>
        <CardInfoStyled isGridView={isGridView}>{published}</CardInfoStyled>
      </CardContentStyled>
    </CardItemStyled>
  );
};

export default memo(CardItem);
