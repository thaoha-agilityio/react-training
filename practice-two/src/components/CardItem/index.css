.card-item {
  width: 282px;
  height: 369px;
  box-shadow: 0px 1px 3px #0000001a;
  border-radius: 5px;
  display: flex;
  flex-direction: column;
}

.card-image {
  width: 249px;
  height: 238px;
  background: var(--color-secondary-100);
  border-radius: 5px;
  display: flex;
  justify-content: center;
  margin: 13px 18px;
}

.card-image img {
  width: 201px;
  height: 202px;
  border-radius: 5px;
  margin-top: 18px;
}

.card-contents {
  margin: 21px 19px;
}

.title {
  font-family: 'Rubik';
  font-weight: 500;
  font-size: 18px;
  color: var(--color-primary-50);
  line-height: calc(21 / 18);
}

.info {
  font-family: 'Noto_Sans';
  font-weight: 500;
  font-size: 13px;
  line-height: calc(18 / 13);
  color: var(--color-secondary-400);
}
