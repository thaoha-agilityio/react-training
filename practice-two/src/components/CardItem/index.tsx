import { memo, useContext } from 'react';

import Avatar from '../Avatar';

import { BooksContext } from '@/contexts/BooksContext';
import { IBook } from '@/types/book';

import './index.css';

interface IProps {
  id: string;
  onShowModal: (id: string) => void;
}

const CardItem = ({ id, onShowModal }: IProps): React.ReactElement => {
  const { getBookById } = useContext(BooksContext);
  const book: IBook = getBookById(id);

  const handleClick = (): void => {
    onShowModal(id);
  };

  return (
    <div className="card-item" data-id={id} onClick={handleClick}>
      <div className="card-image-wrapper">
        <Avatar url={book.avatar} alt={book.name} size="small" />
      </div>
      <div className="card-contents">
        <h2 className="card-title">{book.name}</h2>
        <p className="card-info">{book.author}</p>
        <p className="card-info">{book.published}</p>
      </div>
    </div>
  );
};

export default memo(CardItem);
