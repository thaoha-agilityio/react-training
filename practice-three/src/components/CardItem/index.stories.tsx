import { ComponentStory, ComponentMeta } from "@storybook/react";

import { book } from "@/constants/mockData";

import CardItem from ".";

export default {
  title: "CardItem",
  component: CardItem,
  argTypes: {
    item: {
      defaultValue: book,
    },
    isGridView: {
      defaultValue: true,
      control: { type: "boolean" },
    },
    onSetSelectedBookId: { action: "1" }
  },
} as ComponentMeta<typeof CardItem>;

const Template: ComponentStory<typeof CardItem> = (args) => (
  <CardItem {...args} />
);

export const Default = Template.bind({});
