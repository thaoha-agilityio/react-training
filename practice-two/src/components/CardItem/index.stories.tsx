import { ComponentMeta, ComponentStory } from '@storybook/react';

import CardItem from '.';

export default {
  title: 'CardItem',
  component: CardItem,
} as ComponentMeta<typeof CardItem>;

const Template: ComponentStory<typeof CardItem> = (args) => <CardItem {...args} />;

const Default = Template.bind({});
Default.args = {
  id: '1',
  name: 'Taking Things Quietly',
  avatar: 'http://digital.library.lse.ac.uk/objects/lse:dun949mif/view',
  author: 'Francis ',
  published: '1990',
};

export { Default };
