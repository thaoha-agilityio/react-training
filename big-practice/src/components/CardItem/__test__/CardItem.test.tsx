import { fireEvent, render } from '@testing-library/react';

//  Constants
import { MOCK_PRODUCTS } from '@constants';

import CardItem from '../index';

const mockProps = {
  item: MOCK_PRODUCTS[1],
  onOpen: jest.fn(),
  onEditItem: jest.fn(),
  onShowDetailItem: jest.fn(),
  onAddToCart: jest.fn(),
};

describe('CardItem Component', () => {
  it('renders product information correctly', () => {
    const { container, getByText, getByAltText } = render(<CardItem {...mockProps} />);

    // Assert that the product name, description, and price are displayed
    expect(getByText(MOCK_PRODUCTS[1].name)).toBeInTheDocument();
    expect(getByText(MOCK_PRODUCTS[1].description)).toBeInTheDocument();
    expect(getByText(`Rp ${MOCK_PRODUCTS[1].price}`)).toBeInTheDocument();
    expect(getByAltText('card-item')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should call functions when button is clicked', () => {
    const { getByTestId, getByText } = render(<CardItem {...mockProps} />);

    const deleteBtn = getByTestId('delete-btn');
    fireEvent.click(deleteBtn);
    expect(mockProps.onOpen).toHaveBeenCalled();

    // onEditItem was called
    const editBtn = getByTestId('edit-btn');
    fireEvent.click(editBtn);
    expect(mockProps.onEditItem).toHaveBeenCalled();

    // onShowDetailItem was called
    const detailBtn = getByTestId('detail-btn');
    fireEvent.click(detailBtn);
    expect(mockProps.onShowDetailItem).toHaveBeenCalled();

    // onAddToCart
    const addToCartBtn = getByText('Add to cart');
    fireEvent.click(addToCartBtn);
    expect(mockProps.onAddToCart).toHaveBeenCalled();
  });
});
