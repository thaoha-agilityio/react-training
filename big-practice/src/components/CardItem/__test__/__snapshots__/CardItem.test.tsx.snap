// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardItem Component renders product information correctly 1`] = `
<div>
  <div
    class="css-kjafn5"
  >
    <div
      class="chakra-card css-2dlz7"
    >
      <div
        class="chakra-card__body css-0"
        role="group"
      >
        <div
          class="css-1bl94k9"
        >
          <img
            alt="card-item"
            class="chakra-image__placeholder css-uwwqev"
            src="https://www.charlotteathleticclub.com/assets/camaleon_cms/image-not-found-4a963b95bf081c3ea02923dceaeb3f8085e1a654fc54840aac61a57a60903fef.png"
          />
        </div>
        <div
          class="css-1hyfx7x"
        >
          <div
            class="css-e4qces"
          />
          <div
            class="css-ll7lr2"
          >
            <button
              class="chakra-button css-s3fhhx"
              type="button"
            >
              Add to cart
            </button>
            <div
              class="css-7hqoax"
            >
              <div
                class="css-u0pkxt"
              >
                <button
                  class="chakra-button css-117sk0y"
                  data-testid="delete-btn"
                  type="button"
                >
                  <span
                    class="chakra-button__icon css-1wh2kri"
                  >
                    <svg
                      aria-hidden="true"
                      class="chakra-icon css-onkibi"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <g
                        fill="currentColor"
                      >
                        <path
                          d="M19.452 7.5H4.547a.5.5 0 00-.5.545l1.287 14.136A2 2 0 007.326 24h9.347a2 2 0 001.992-1.819L19.95 8.045a.5.5 0 00-.129-.382.5.5 0 00-.369-.163zm-9.2 13a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zm5 0a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zM22 4h-4.75a.25.25 0 01-.25-.25V2.5A2.5 2.5 0 0014.5 0h-5A2.5 2.5 0 007 2.5v1.25a.25.25 0 01-.25.25H2a1 1 0 000 2h20a1 1 0 000-2zM9 3.75V2.5a.5.5 0 01.5-.5h5a.5.5 0 01.5.5v1.25a.25.25 0 01-.25.25h-5.5A.25.25 0 019 3.75z"
                        />
                      </g>
                    </svg>
                  </span>
                  Delete
                </button>
                <button
                  class="chakra-button css-117sk0y"
                  data-testid="edit-btn"
                  type="button"
                >
                  <span
                    class="chakra-button__icon css-1wh2kri"
                  >
                    <svg
                      aria-hidden="true"
                      class="chakra-icon css-onkibi"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <g
                        fill="none"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-width="2"
                      >
                        <path
                          d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
                        />
                        <path
                          d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
                        />
                      </g>
                    </svg>
                  </span>
                  Edit
                </button>
                <button
                  class="chakra-button css-117sk0y"
                  data-testid="detail-btn"
                  type="button"
                >
                  <span
                    class="chakra-button__icon css-1wh2kri"
                  >
                    <svg
                      aria-hidden="true"
                      class="chakra-icon css-onkibi"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                  Detail
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="chakra-card__footer css-k008qs"
      >
        <div
          class="chakra-stack css-1272ag6"
        >
          <h2
            class="chakra-heading css-hwp7z0"
          >
            name 1
          </h2>
          <p
            class="chakra-text css-szn3bb"
          >
            description
          </p>
          <p
            class="chakra-text css-3ut35e"
          >
            Rp 
            8
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
